File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root: return
        
        currentNode = root

        while currentNode:
            if currentNode.left:
                last = currentNode.left
                while last.right:
                    last = last.right
                last.right = currentNode.right
                currentNode.right = currentNode.left
                currentNode.left = None
            
            currentNode = currentNode.right
        
        return currentNode
                

File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;

        TreeNode currentNode = root;
        while (currentNode != null) {
            if (currentNode.left != null) {
                TreeNode last = currentNode.left;
                while (last.right != null) {
                    last = last.right;
                }

                last.right = currentNode.right;
                currentNode.right = currentNode.left;
                currentNode.left = null;
            }

            currentNode = currentNode.right;
        }
        
    }
}